File: agent_inspect/importer.py
"""外部 span 导出导入:把遵循 OpenInference 语义约定的 span 树映射为决策点 trace。

只读映射,不碰拦截器与记录路径;落库复用 recorder.persist(增量快照 + 大对象去重),
导入产物与自录 trace 同构——查看 / Fork / 分支 diff 零改动复用。
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Optional

from . import _models as m
from ._server.store.queries import Store
from .interceptor.base import _hash_of
from .recorder import Recorder

KIND_ATTR = "openinference.span.kind"
SPAN_KIND_LLM = "LLM"
SPAN_KIND_TOOL = "TOOL"


class TraceImportError(Exception):
    """导入校验失败(非 JSON 对象 / 缺 span 树 / 无可映射 span)。不产生任何写入。"""


@dataclass
class ImportResult:
    trace_id: str
    root_branch_id: str
    decision_points: int
    skipped: int


def import_trace(store: Store, recorder: Recorder, payload: Any) -> ImportResult:
    """导入一份 span 导出 JSON,返回导入统计。失败抛 TraceImportError,不落库。

    接受两种形态:OTLP JSON 信封(resourceSpans[].scopeSpans[].spans[])或
    扁平 span 列表({"spans": [...]},属性为 {key: value} 对象)。
    """
    spans, agent_name = _extract_spans(payload)
    ordered = _order_spans(spans)
    mapped = [s for s in ordered if _span_kind(s) in (SPAN_KIND_LLM, SPAN_KIND_TOOL)]
    skipped = len(ordered) - len(mapped)
    if not mapped:
        raise TraceImportError(
            "no importable spans found: need spans with "
            f"{KIND_ATTR} in ({SPAN_KIND_LLM}, {SPAN_KIND_TOOL})"
        )

    starts = [_start_sec(s) for s in ordered if _start_sec(s) > 0]
    started_at = min(starts) if starts else m.now()

    trace, branch = store.create_trace_with_root(agent_name, parent_trace_id=None)
    prev_dp_id: Optional[str] = None
    count = 0
    for span in mapped:
        dp = _dp_from_span(trace.id, branch.id, count, span, prev_dp_id)
        if dp.output is not None:
            dp.output_hash = _hash_of(dp.output)
        recorder.persist(dp)
        prev_dp_id = dp.id
        count += 1

    store.set_trace_lifecycle(trace.id, m.LIFECYCLE_DONE)
    store.set_trace_started_at(trace.id, started_at)
    return ImportResult(
        trace_id=trace.id, root_branch_id=branch.id, decision_points=count, skipped=skipped
    )


# ---------------------------------------------------------------------------
# 输入形态:OTLP 信封 / 扁平列表
# ---------------------------------------------------------------------------
def _extract_spans(payload: Any) -> tuple[list[dict], str]:
    if not isinstance(payload, dict):
        raise TraceImportError("export must be a JSON object")
    agent_name = "imported"
    spans: list[dict] = []
    resource_spans = payload.get("resourceSpans")
    if isinstance(resource_spans, list):
        for res in resource_spans:
            if not isinstance(res, dict):
                continue
            name = _resource_service_name(res)
            if name:
                agent_name = name
            for scope in res.get("scopeSpans") or []:
                if isinstance(scope, dict):
                    spans.extend(s for s in scope.get("spans") or [] if isinstance(s, dict))
    elif isinstance(payload.get("spans"), list):
        spans = [s for s in payload["spans"] if isinstance(s, dict)]
        agent_name = payload.get("agent_name") or payload.get("service.name") or agent_name
    if not spans:
        raise TraceImportError(
            "no spans found in export (need resourceSpans[].scopeSpans[].spans[] or spans[])"
        )
    return spans, str(agent_name)


def _resource_service_name(res: dict) -> Optional[str]:
    attrs = res.get("resource", {}).get("attributes")
    flat = _flatten_otlp_attrs(attrs) if isinstance(attrs, list) else (attrs or {})
    name = flat.get("service.name")
    return str(name) if name else None


# ---------------------------------------------------------------------------
# span 字段读取(OTLP camelCase / 扁平 snake_case 兼容)
# ---------------------------------------------------------------------------
def _attr_map(span: dict) -> dict:
    attrs = span.get("attributes")
    if isinstance(attrs, list):
        return _flatten_otlp_attrs(attrs)
    if isinstance(attrs, dict):
        return attrs
    return {}


def _flatten_otlp_attrs(attrs: list) -> dict:
    """OTel 属性数组 [{key, value:{stringValue|...}}] → {key: python_value}。"""
    out: dict = {}
    for a in attrs or []:
        if not isinstance(a, dict) or not a.get("key"):
            continue
        out[a["key"]] = _otlp_value(a.get("value"))
    return out


def _otlp_value(v: Any) -> Any:
    if not isinstance(v, dict):
        return v
    for k in ("stringValue", "intValue", "doubleValue", "boolValue"):
        if k in v:
            return v[k]
    if "arrayValue" in v and isinstance(v["arrayValue"], dict):
        return [_otlp_value(x) for x in v["arrayValue"].get("values") or []]
    if "kvlistValue" in v and isinstance(v["kvlistValue"], dict):
        return _flatten_otlp_attrs(v["kvlistValue"].get("values") or [])
    return v


def _span_id(span: dict) -> str:
    sid = span.get("spanId") or span.get("span_id") or ""
    return str(sid)


def _parent_id(span: dict) -> str:
    pid = span.get("parentSpanId") or span.get("parent_span_id") or ""
    return str(pid)


def _raw_start(span: dict) -> float:
    v = span.get("startTimeUnixNano")
    if v is None:
        v = span.get("start_time")
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def _start_sec(span: dict) -> float:
    """起始时间归一到秒(自动识别 ns / µs / ms / s 数量级)。"""
    n = _raw_start(span)
    return n / _unit_divisor(n)


def _unit_divisor(n: float) -> float:
    if n >= 1e17:
        return 1e9  # ns
    if n >= 1e14:
        return 1e6  # µs
    if n >= 1e11:
        return 1e3  # ms
    return 1.0  # s


def _duration_ms(span: dict) -> Optional[float]:
    e = span.get("endTimeUnixNano")
    if e is None:
        e = span.get("end_time")
    try:
        end = float(e)
    except (TypeError, ValueError):
        return None
    start = _raw_start(span)
    if start <= 0 or end <= start:
        return None
    return round((end - start) / _unit_divisor(start) * 1000, 1)


def _span_kind(span: dict) -> Optional[str]:
    k = _attr_map(span).get(KIND_ATTR)
    return k if isinstance(k, str) else None


# ---------------------------------------------------------------------------
# 排序:父 span 建树 + 深度优先,稳序(start, span_id);孤儿/环兜底为根
# ---------------------------------------------------------------------------
def _order_spans(spans: list[dict]) -> list[dict]:
    by_id = {s_id: s for s in spans if (s_id := _span_id(s))}
    children: dict[str, list[dict]] = {}
    roots: list[dict] = []
    for s in spans:
        pid = _parent_id(s)
        if pid and pid in by_id and pid != _span_id(s):
            children.setdefault(pid, []).append(s)
        else:
            roots.append(s)
    order_key = lambda s: (_raw_start(s), _span_id(s))  # noqa: E731
    roots.sort(key=order_key)
    for lst in children.values():
        lst.sort(key=order_key)
    out: list[dict] = []
    seen: set[int] = set()

    def visit(s: dict) -> None:
        if id(s) in seen:
            return
        seen.add(id(s))
        out.append(s)
        for c in children.get(_span_id(s), []):
            visit(c)

    for r in roots:
        visit(r)
    for s in sorted(spans, key=order_key):  # 环引用兜底:未被树覆盖的按全局序补齐
        visit(s)
    return out


# ---------------------------------------------------------------------------
# 映射:span → 决策点(形态与插桩器对齐,Fork 回放的前提)
# ---------------------------------------------------------------------------
def _dp_from_span(
    trace_id: str, branch_id: str, step: int, span: dict, prev_dp_id: Optional[str]
) -> m.DecisionPoint:
    kind_attr = _span_kind(span)
    attrs = _attr_map(span)
    name = span.get("name") or "span"
    if kind_attr == SPAN_KIND_LLM:
        dp_kind, agent_id = m.KIND_LLM, str(name)
        input_context = _llm_input(attrs)
        output = _llm_output(attrs)
    else:
        dp_kind, agent_id = m.KIND_TOOL, str(attrs.get("tool.name") or name)
        input_context = _tool_input(attrs, str(name))
        output = _tool_output(attrs)
    meta: dict = {"imported": True}
    if _span_id(span):
        meta["imported_span_id"] = _span_id(span)
    lat = _duration_ms(span)
    if lat is not None:
        meta["latency_ms"] = lat
    return m.DecisionPoint(
        id=m.new_id("dp"),
        trace_id=trace_id,
        branch_id=branch_id,
        step_index=step,
        kind=dp_kind,
        agent_id=agent_id,
        input_context=input_context,
        output=output,
        cause_edge=[prev_dp_id] if prev_dp_id else [],
        meta=meta,
    )


def _llm_input(attrs: dict) -> dict:
    return {
        "messages": _messages_from(attrs, "llm.input_messages"),
        "model": attrs.get("llm.model_name") or attrs.get("llm.model"),
        "params": _json_dict(attrs.get("llm.invocation_parameters")),
    }


def _llm_output(attrs: dict) -> dict:
    """输出对齐 _shape_llm / _shape_response:{"content", "tool_calls"}。"""
    msgs = _messages_from(attrs, "llm.output_messages")
    content, tool_calls = None, []
    for msg in msgs:
        if (msg.get("role") or "assistant") == "assistant":
            content = msg.get("content")
            tool_calls = msg.get("tool_calls") or []
            break
    if content is None and msgs:
        content = msgs[-1].get("content")
    return {"content": content, "tool_calls": tool_calls or _json_list(attrs.get("llm.tool_calls"))}


def _tool_input(attrs: dict, span_name: str) -> dict:
    return {
        "tool": attrs.get("tool.name") or span_name,
        "args": _json_any(attrs.get("tool.parameters")),
    }


def _tool_output(attrs: dict) -> dict:
    """输出对齐 Serializer.tool_output / _reconstruct_tool:{"result", "is_error"}。"""
    return {
        "result": _json_any(attrs.get("tool.return_value")),
        "is_error": bool(attrs.get("tool.is_error")),
    }


# ---------------------------------------------------------------------------
# 属性值规整:JSON 字符串二次解析(失败降级为原文本)
# ---------------------------------------------------------------------------
def _json_any(v: Any) -> Any:
    if isinstance(v, str):
        try:
            return json.loads(v)
        except (TypeError, ValueError):
            return v
    return v


def _json_dict(v: Any) -> dict:
    d = _json_any(v)
    return d if isinstance(d, dict) else {}


def _json_list(v: Any) -> list:
    d = _json_any(v)
    return d if isinstance(d, list) else []


def _messages_from(attrs: dict, prefix: str) -> list[dict]:
    """读取消息列表,支持两种 OpenInference 编码:

    1. 整体属性(JSON 字符串或已解析列表),元素形如 {"message": {role, content, tool_calls}} 或 {role, content};
    2. 扁平点分键:`prefix.<i>.message.role` / `prefix.<i>.message.content`。
    """
    raw = _json_list(attrs.get(prefix))
    if raw:
        return [_msg_record(x) for x in raw]
    per_idx: dict[int, dict] = {}
    for k, v in attrs.items():
        if not isinstance(k, str) or not k.startswith(prefix + "."):
            continue
        parts = k[len(prefix) + 1 :].split(".")
        if not parts or not parts[0].isdigit():
            continue
        slot = per_idx.setdefault(int(parts[0]), {})
        slot[".".join(parts[1:])] = v
    out = []
    for i in sorted(per_idx):
        slot = per_idx[i]
        rec: dict = {
            "role": slot.get("message.role", "user"),
            "content": slot.get("message.content"),
        }
        if slot.get("message.tool_calls"):  # 部分导出把 tool_calls 整体挂在扁平键上
            rec["tool_calls"] = slot["message.tool_calls"]
        out.append(rec)
    return out


def _msg_record(msg: Any) -> dict:
    role = _msg_field(msg, "role")
    content = _msg_field(msg, "content")
    tool_calls = _msg_field(msg, "tool_calls")
    rec: dict = {"role": role, "content": content}
    if tool_calls:  # assistant 的工具调用透传,回放时 reconstruct 需要
        rec["tool_calls"] = tool_calls
    return rec


def _msg_field(msg: Any, field: str) -> Any:
    if isinstance(msg, dict):
        inner = msg.get("message")
        if isinstance(inner, dict):
            return inner.get(field)
        return msg.get(field) or msg.get(f"message.{field}")
    return None

File: sdks/node/src/importer.ts
// 外部 span 导出导入:与 Python importer.py 同语义(两形态、映射、拒绝不落库)。
import type { DecisionPoint } from "./models.js";
import { newId, now } from "./models.js";
import type { Store } from "./store.js";

export const KIND_ATTR = "openinference.span.kind";
export const SPAN_KIND_LLM = "LLM";
export const SPAN_KIND_TOOL = "TOOL";

export class TraceImportError extends Error {}

export interface ImportResult {
  traceId: string;
  rootBranchId: string;
  decisionPoints: number;
  skipped: number;
}

// ---------------------------------------------------------------------------
// 输入形态:OTLP 信封 / 扁平列表
// ---------------------------------------------------------------------------
function extractSpans(payload: unknown): { spans: Record<string, unknown>[]; agentName: string } {
  if (!payload || typeof payload !== "object") {
    throw new TraceImportError("export must be a JSON object");
  }
  const p = payload as Record<string, any>;
  let agentName = "imported";
  let spans: Record<string, unknown>[] = [];

  if (Array.isArray(p.resourceSpans)) {
    for (const res of p.resourceSpans) {
      if (!res || typeof res !== "object") continue;
      const resAttrs = res.resource?.attributes;
      const svc = (Array.isArray(resAttrs) ? flatAttrs(resAttrs) : resAttrs ?? {})["service.name"];
      if (svc) agentName = String(svc);
      for (const scope of res.scopeSpans ?? []) {
        if (scope && typeof scope === "object") {
          spans.push(...((scope.spans ?? []) as any[]).filter((s) => s && typeof s === "object"));
        }
      }
    }
  } else if (Array.isArray(p.spans)) {
    spans = p.spans.filter((s) => s && typeof s === "object");
    agentName = p.agent_name || p["service.name"] || agentName;
  }
  if (!spans.length) {
    throw new TraceImportError(
      "no spans found in export (need resourceSpans[].scopeSpans[].spans[] or spans[])",
    );
  }
  return { spans, agentName: String(agentName) };
}

// ---------------------------------------------------------------------------
// span 字段读取(OTLP camelCase / 扁平 snake_case 兼容)
// ---------------------------------------------------------------------------
function flatAttrs(values: unknown[]): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const a of values) {
    const a2 = a as { key?: string; value?: unknown };
    if (a2 && a2.key) out[a2.key] = otlpValue(a2.value);
  }
  return out;
}

function attrMap(span: Record<string, any>): Record<string, unknown> {
  const attrs = span.attributes;
  if (Array.isArray(attrs)) {
    const out: Record<string, unknown> = {};
    for (const a of attrs) {
      if (a && a.key) out[a.key] = otlpValue(a.value);
    }
    return out;
  }
  return attrs && typeof attrs === "object" ? attrs : {};
}

function otlpValue(v: any): unknown {
  if (!v || typeof v !== "object") return v;
  for (const k of ["stringValue", "intValue", "doubleValue", "boolValue"]) {
    if (k in v) return v[k];
  }
  if (v.arrayValue && typeof v.arrayValue === "object") {
    return (v.arrayValue.values ?? []).map(otlpValue);
  }
  if (v.kvlistValue && typeof v.kvlistValue === "object") {
    return flatAttrs(v.kvlistValue.values ?? []);
  }
  return v;
}

function spanId(span: Record<string, any>): string {
  return String(span.spanId ?? span.span_id ?? "");
}

function parentId(span: Record<string, any>): string {
  return String(span.parentSpanId ?? span.parent_span_id ?? "");
}

function rawStart(span: Record<string, any>): number {
  const v = span.startTimeUnixNano ?? span.start_time;
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
}

function unitDivisor(n: number): number {
  if (n >= 1e17) return 1e9; // ns
  if (n >= 1e14) return 1e6; // µs
  if (n >= 1e11) return 1e3; // ms
  return 1;
}

function startSec(span: Record<string, any>): number {
  return rawStart(span) / unitDivisor(rawStart(span));
}

function durationMs(span: Record<string, any>): number | null {
  const endV = span.endTimeUnixNano ?? span.end_time;
  const end = Number(endV);
  const start = rawStart(span);
  if (!Number.isFinite(end) || start <= 0 || end <= start) return null;
  return Math.round(((end - start) / unitDivisor(start)) * 1000 * 10) / 10;
}

function spanKind(span: Record<string, any>): string | null {
  const k = attrMap(span)[KIND_ATTR];
  return typeof k === "string" ? k : null;
}

// ---------------------------------------------------------------------------
// 排序:父 span 建树 + DFS(稳序 start/spanId);孤儿/环兜底
// ---------------------------------------------------------------------------
function orderSpans(spans: Record<string, any>[]): Record<string, any>[] {
  const byId = new Map<string, Record<string, any>>();
  for (const s of spans) {
    const id = spanId(s);
    if (id) byId.set(id, s);
  }
  const children = new Map<string, Record<string, any>[]>();
  const roots: Record<string, any>[] = [];
  for (const s of spans) {
    const pid = parentId(s);
    const sid = spanId(s);
    if (pid && byId.has(pid) && pid !== sid) {
      if (!children.has(pid)) children.set(pid, []);
      children.get(pid)!.push(s);
    } else {
      roots.push(s);
    }
  }
  const key = (s: Record<string, any>) => [rawStart(s), spanId(s)] as const;
  roots.sort((a, b) => {
    const ka = key(a);
    const kb = key(b);
    return ka[0] - kb[0] || (ka[1] < kb[1] ? -1 : 1);
  });
  for (const list of children.values()) {
    list.sort((a, b) => {
      const ka = key(a);
      const kb = key(b);
      return ka[0] - kb[0] || (ka[1] < kb[1] ? -1 : 1);
    });
  }
  const out: Record<string, any>[] = [];
  const seen = new Set<object>();
  const visit = (s: Record<string, any>) => {
    if (seen.has(s)) return;
    seen.add(s);
    out.push(s);
    for (const c of children.get(spanId(s)) ?? []) visit(c);
  };
  for (const r of roots) visit(r);
  for (const s of [...spans].sort((a, b) => {
    const ka = key(a);
    const kb = key(b);
    return ka[0] - kb[0] || (ka[1] < kb[1] ? -1 : 1);
  })) {
    visit(s);
  }
  return out;
}

// ---------------------------------------------------------------------------
// 映射:span → 决策点(与 Node 插桩器 shape 同形)
// ---------------------------------------------------------------------------
function jsonAny(v: unknown): unknown {
  if (typeof v === "string") {
    try {
      return JSON.parse(v);
    } catch {
      return v;
    }
  }
  return v;
}

function jsonDict(v: unknown): Record<string, unknown> {
  const d = jsonAny(v);
  return d && typeof d === "object" && !Array.isArray(d) ? (d as Record<string, unknown>) : {};
}

function jsonList(v: unknown): unknown[] {
  const d = jsonAny(v);
  return Array.isArray(d) ? d : [];
}

function msgField(msg: unknown, field: string): unknown {
  if (msg && typeof msg === "object") {
    const inner = (msg as any).message;
    if (inner && typeof inner === "object") return inner[field];
    return (msg as any)[field] ?? (msg as any)[`message.${field}`];
  }
  return null;
}

function msgRecord(msg: unknown): Record<string, unknown> {
  const rec: Record<string, unknown> = {
    role: msgField(msg, "role") ?? null,
    content: msgField(msg, "content") ?? null,
  };
  const tc = msgField(msg, "tool_calls");
  if (tc) rec["tool_calls"] = tc;
  return rec;
}

function messagesFrom(attrs: Record<string, unknown>, prefix: string): Record<string, unknown>[] {
  const raw = jsonList(attrs[prefix]);
  if (raw.length) return raw.map(msgRecord);
  const perIdx = new Map<number, Record<string, unknown>>();
  for (const [k, v] of Object.entries(attrs)) {
    if (!k.startsWith(prefix + ".")) continue;
    const parts = k.slice(prefix.length + 1).split(".");
    if (!parts[0] || !/^\d+$/.test(parts[0])) continue;
    const idxKey = Number(parts[0]);
    const slot = perIdx.get(idxKey) ?? {};
    perIdx.set(idxKey, slot);
    slot[parts.slice(1).join(".")] = v;
  }
  const out: Record<string, unknown>[] = [];
  for (const i of [...perIdx.keys()].sort((a, b) => a - b)) {
    const slot = perIdx.get(i)!;
    const rec: Record<string, unknown> = {
      role: slot["message.role"] ?? "user",
      content: slot["message.content"] ?? null,
    };
    if (slot["message.tool_calls"]) rec["tool_calls"] = slot["message.tool_calls"];
    out.push(rec);
  }
  return out;
}

function dpFromSpan(
  traceId: string,
  branchId: string,
  step: number,
  span: Record<string, any>,
  prevDpId: string | null,
): DecisionPoint {
  const kindAttr = spanKind(span);
  const attrs = attrMap(span);
  const name = (span.name as string) || "span";
  let dpKind: "llm" | "tool";
  let agentId: string;
  let inputContext: Record<string, unknown>;
  let output: Record<string, unknown>;

  if (kindAttr === SPAN_KIND_LLM) {
    dpKind = "llm";
    agentId = name;
        inputContext = {
      messages: messagesFrom(attrs, "llm.input_messages"),
      model: attrs["llm.model_name"] || attrs["llm.model"] || null,
      params: jsonDict(attrs["llm.invocation_parameters"]),
    } as Record<string, unknown>;
    const outMsgs = messagesFrom(attrs, "llm.output_messages");
    let content: unknown = null;
    let toolCalls: unknown[] = [];
    for (const m of outMsgs) {
      if ((m["role"] ?? "assistant") === "assistant") {
        content = m["content"];
        toolCalls = Array.isArray(m["tool_calls"]) ? (m["tool_calls"] as unknown[]) : [];
        break;
      }
    }
    if (content === null && outMsgs.length) content = outMsgs[outMsgs.length - 1]["content"];
    if (!toolCalls.length) toolCalls = jsonList(attrs["llm.tool_calls"]);
    output = { content, tool_calls: tool_calls_ensure(toolCalls) };
  } else {
    dpKind = "tool";
    agentId = String(attrs["tool.name"] ?? name);
    inputContext = {
      tool: attrs["tool.name"] ?? name,
      args: jsonAny(attrs["tool.parameters"]),
    };
    output = {
      result: jsonAny(attrs["tool.return_value"]),
      is_error: Boolean(attrs["tool.is_error"]),
    };
  }

  const meta: Record<string, unknown> = { imported: true };
  const sid = spanId(span);
  if (sid) meta["imported_span_id"] = sid;
  const lat = durationMs(span);
  if (lat !== null) meta["latency_ms"] = lat;

  return {
    id: newId("dp"),
    trace_id: traceId,
    branch_id: branchId,
    step_index: step,
    kind: dpKind,
    agent_id: agentId,
    input_context: inputContext,
    output,
    output_hash: null,
    cause_edge: prevDpId ? [prevDpId] : [],
    meta,
  };
}

function tool_calls_ensure(tc: unknown): unknown[] {
  return Array.isArray(tc) ? tc : [];
}

// ---------------------------------------------------------------------------
// 入口
// ---------------------------------------------------------------------------
export function importTrace(store: Store, payload: unknown): ImportResult {
  const { spans, agentName } = extractSpans(payload);
  const ordered = orderSpans(spans);
  const mapped = ordered.filter((s) => [SPAN_KIND_LLM, SPAN_KIND_TOOL].includes(spanKind(s) ?? ""));
  const skipped = ordered.length - mapped.length;
  if (!mapped.length) {
    throw new TraceImportError(
      `no importable spans found: need spans with ${KIND_ATTR} in (${SPAN_KIND_LLM}, ${SPAN_KIND_TOOL})`,
    );
  }

  const starts = ordered.map(startSec).filter((s) => s > 0);
  const startedAt = starts.length ? Math.min(...starts) : now();

  const { trace, branch } = store.createTraceWithRoot(agentName);
  store.setTraceLifecycle(trace.id, "done");
  store.setTraceStartedAt(trace.id, startedAt);

  let prevDpId: string | null = null;
  let count = 0;
  for (const span of mapped) {
    const dp = dpFromSpan(trace.id, branch.id, count, span, prevDpId);
    store.writeDecisionPoint(dp);
    prevDpId = dp.id;
    count += 1;
  }
  return {
    traceId: trace.id,
    rootBranchId: branch.id,
    decisionPoints: count,
    skipped,
  };
}
